logical/framework/openapi.go

```diff
@@ -0,0 +1,413 @@
+		replacedPaths = append(replacedPaths, path)
+	}
+
+	return replacedPaths
+}
+
+// schemaType is a subset of the JSON Schema elements used as a target
+// for conversions from Vault's standard FieldTypes.
+type schemaType struct {
+	baseType string
+	items    string
+	format   string
+}
+
+// convertType translates a FieldType into an OpenAPI type.
+// In the case of arrays, a subtype is returned as well.
+func convertType(t FieldType) schemaType {
+	ret := schemaType{}
+
+	switch t {
+	case TypeString, TypeNameString, TypeLowerCaseString, TypeHeader:
+		ret.baseType = "string"
+	case TypeInt:
+		ret.baseType = "number"
+	case TypeDurationSecond:
+		ret.baseType = "number"
+		ret.format = "seconds"
+	case TypeBool:
+		ret.baseType = "boolean"
+	case TypeMap:
+		ret.baseType = "object"
+		ret.format = "map"
+	case TypeKVPairs:
+		ret.baseType = "object"
+		ret.format = "kvpairs"
+	case TypeSlice, TypeStringSlice, TypeCommaStringSlice:
+		ret.baseType = "array"
+		ret.items = "string"
+	case TypeCommaIntSlice:
+		ret.baseType = "array"
+		ret.items = "number"
+	default:
+		log.L().Warn("error parsing field type", "type", t)
+		ret.format = "unknown"
+	}
+
+	return ret
+}
+
+// cleanString prepares s for inclusion in the output
+func cleanString(s string) string {
+	// clean leading/trailing whitespace, and replace whitespace runs into a single space
+	s = strings.TrimSpace(s)
+	s = wsRe.ReplaceAllString(s, " ")
+	return s
+}
+
+// splitFields partitions fields into path and body groups
+// The input pattern is expected to have been run through expandPattern,
+// with paths parameters denotes in {braces}.
+func splitFields(allFields map[string]*FieldSchema, pattern string) (pathFields, bodyFields map[string]*FieldSchema) {
+	pathFields = make(map[string]*FieldSchema)
+	bodyFields = make(map[string]*FieldSchema)
+
+	for _, match := range pathFieldsRe.FindAllStringSubmatch(pattern, -1) {
+		name := match[1]
+		pathFields[name] = allFields[name]
+	}
+
+	for name, field := range allFields {
+		if _, ok := pathFields[name]; !ok {
+			// Header fields are in "parameters" with other path fields
+			if field.Type == TypeHeader {
+				pathFields[name] = field
+			} else {
+				bodyFields[name] = field
+			}
+		}
+	}
+
+	return pathFields, bodyFields
+}
+
+// cleanedResponse is identical to logical.Response but with nulls
+// removed from from JSON encoding
+type cleanedResponse struct {
+	Secret   *logical.Secret            `json:"secret,omitempty"`
+	Auth     *logical.Auth              `json:"auth,omitempty"`
+	Data     map[string]interface{}     `json:"data,omitempty"`
+	Redirect string                     `json:"redirect,omitempty"`
+	Warnings []string                   `json:"warnings,omitempty"`
+	WrapInfo *wrapping.ResponseWrapInfo `json:"wrap_info,omitempty"`
+}
+
+func cleanResponse(resp *logical.Response) *cleanedResponse {
+	var r cleanedResponse
+
+	if mapstructure.Decode(resp, &r) != nil {
+		return nil
```

Comment:
In the case that an error is returned from `Decode`, would it be wise to log it and the response here before returning nil? Maybe at a debug or warn level?

Reply:
👍 I've changed it to return an error parameter too.